### PR TITLE
Allow atm processes to compute tendencies of updated fields

### PR DESCRIPTION
**Pull request by bartgol**
Upon request, an atm proc can compute tendencies of any of its updated fields, which can later be used in IO. Some notes below.

- The field must be updated (that is, both required and computed by the atm proc), to ensure tendency is always well formed (including at first time step).
- For field X, the computed tendency will have name `$procName_X_tend`, to allow multiple atm proc to compute their tendency for the same field.
- To request tendency, set `compute_tendencies` entry in atm proc settings to a list of desired fields, such as `compute_tendencies: [T_mid, qc]`.
- If an atm proc operates on 2+ grid and the same field appears on more than 1 grid, you must specify the grid of the field: `compute_tendencies: [T_mid#Dynamics, qc#Physics]`. This notation is ugly, but it's the only one that is highly unlikely to cause name clashes. For these fields, the tendency name is `$procName_X#$gridName_tend`. Again, quite ugly, but I also suspect we will very rarely have an atm proc acting on 2+ grids, so this will likely be a never-hit corner case anyways.
- Tendencies are not available for groups (e.g., cannot ask for `TRACERS_tend`). If you need the tendency of a group, you must list all the members by hand. Supporting groups was a bit involved in some tracking stuff.
- All tendencies are automatically available in the model field manager(s), and can therefore be added to IO field lists.
- Tendencies are calculated via the newly added `update` method of `Field`.

TODO:
- add unit test that saves both X and X_tend, and verifies that, X_tend(k) = (X(k) - X(k-1)) / dt, for k=1,...,N.
- add a tendency calculation to a CIME test

### CODE REVIEW

File: components/eamxx/src/share/field/field_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+      " - x/y data type  : " + e2str(dt_x) + "\n"
+      " - coeff data type: " + e2str(dt_st) + "\n");
+
+  if (dt_x==DataType::IntType) {
+    return update_impl<HD,int>(x,alpha,beta);
+  } else if (dt_x==DataType::FloatType) {
+    return update_impl<HD,float>(x,alpha,beta);
+  } else if (dt_x==DataType::DoubleType) {
+    return update_impl<HD,double>(x,alpha,beta);
+  }
```
Comment (tcclevenger) [inline components/eamxx/src/share/field/field_impl.hpp:392]: Currently a Field requires one of these 3 data types, but could this change in the future? If a new DT is added without updating this function, `update()` would do nothing (but silently) since `update_impl()` just would not be called. Should we `else { Error("invalid data type") }` here? Then if a new DT is added, anyone trying to use this function would at least be tipped off that something is wrong.

**Review by AaronDonahue — COMMENTED** [review]
Looks great! I had a few notes on stylistic things, nothing major.  I do have a question about the PR in general, will this PR also make it possible to transition to tendency updates in processes for actual state updates?  I'm thinking forward to when we will eventually add parallel-splitting and will need processes to produce tendencies for all "updated" variables rather than explicitly change their states.

**Comment by PeterCaldwell** [comment]
Thanks Luca, this looks great. A couple comments on your first comment on this PR:
1. Am I correct that the way a user will set up a run to output Tmid from P3 (for example) is that they will 1). add a `compute_tendencies` entry in the input yaml block for that process AND 2). they will request P3_Tmid_tend in one of the output yaml files? It would be a bit slicker if the user only needed to specify in the output stream that they wanted a variable. I suspect this would require saving all output tendencies all the time just in case someone wanted one - which is lame - and would also be a lot of work - which isn't worth it - but thought I'd ask in case it's easy. 
2. Is dynamics an example of a process that could have output on either the dyn or the physics grid? You say we'll probably never hit this edge case, but won't we hit it every time we ask for something from dynamics?
3. It's totally fine for only updated fields and non-group variables to be available for _tend treatment, but we should add error messages if someone asks for an inappropriate or non-existent field. Maybe you already have this - I haven't checked. 
5. a use case for this code will be enabling parallel splitting of processes, where we need each process to return its tendencies and all the tendencies for all the processes will be added together in order to do the timestepping. I think everything you've done will be fine for that and just mention it in case it triggers some ideas on your part.

Basically, these are just thoughts rather than anything actionable. I think the PR looks great and the "ugly" formatting seems fine and appropriate to me.

File: components/eamxx/src/share/field/field_impl.hpp (hunk 1 above)
Comment (bartgol) [reply components/eamxx/src/share/field/field_impl.hpp]: Good catch!

File: components/eamxx/src/share/atm_process/atmosphere_process.cpp [hunk 2]
```diff
@@ -93,8 +93,33 @@ void AtmosphereProcess::run (const double dt) {
       compute_column_conservation_checks_data(dt_sub);
     }
 
+    // Init single step tendencies (if any) with current value of output field
+    for (auto& it : m_proc_tendencies_single_step) {
+      const auto& tname = it.first;
+      const auto& fname = m_tend_to_field.at(tname);
+      const auto& f     = get_field_out(fname);
+      // auto f_out = get_out_field_or_group(it.first,it.second);
```
[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/src/share/atm_process/atmosphere_process.cpp]: Nope, I removed it.

**Comment by bartgol** [comment]
> 1. Am I correct that the way a user will set up a run to output Tmid from P3 (for example) is that they will 1). add a `compute_tendencies` entry in the input yaml block for that process AND 2). they will request P3_Tmid_tend in one of the output yaml files? It would be a bit slicker if the user only needed to specify in the output stream that they wanted a variable. I suspect this would require saving all output tendencies all the time just in case someone wanted one - which is lame - and would also be a lot of work - which isn't worth it - but thought I'd ask in case it's easy.

That's correct. We don't want to always compute tendencies (which, to be precise, is more of a memory burden than a computational one, given that we require 2 extra field for each updated field for each atm proc). In order to be slick, we would have to parse the output file specification _before_ we allocate the fields in the FM, which kind of turns the whole flow upside down. I'm gonna say that modifying two files is not too big of a burden.

> 2. Is dynamics an example of a process that could have output on either the dyn or the physics grid? You say we'll probably never hit this edge case, but won't we hit it every time we ask for something from dynamics?
Well, yes and no. The dyn-grid fields in dynamics are "internal" fields, the AD only needs to know about them b/c of restart reasons. They _are_ in the FM (for restarts, mainly), but are not "updated" fields, in the sense that they are not listed as required/computed fields. To see what fields are required/computed, you just need to look for `add_field<..>(...)` calls inside the atm proc interface. Those are the fields we think of. So in this sense, as of today, no atm proc requires/computes a field on a grid other than physics.

> 3. It's totally fine for only updated fields and non-group variables to be available for _tend treatment, but we should add error messages if someone asks for an inappropriate or non-existent field. Maybe you already have this - I haven't checked.

Yes, I did add that check [here](https://github.com/E3SM-Project/scream/pull/2216/files#diff-32ff93d533c1db4e010d3d11e224b2976c3bcf26b173ba5a1f9c636daee585d4R232).

> 4. a use case for this code will be enabling parallel splitting of processes, where we need each process to return its tendencies and all the tendencies for all the processes will be added together in order to do the timestepping. I think everything you've done will be fine for that and just mention it in case it triggers some ideas on your part.

Well, I don't know. This is a rough implementation that pays not much attention to the computational/memory costs, since it was envisioned for debug purposes. As I said, computing tendencies adds 2 extra fields per updated field per atm process. I'm sure this won't play nicely at UHR. I assume some infrastructure _can_ be retained for || splitting, but I suspect we will have to adjust it a bit.

> Basically, these are just thoughts rather than anything actionable. I think the PR looks great and the "ugly" formatting seems fine and appropriate to me.
Great, thanks!

File: components/eamxx/src/diagnostics/field_at_level.hpp [hunk 3]
```diff
@@ -3,6 +3,8 @@
 
 #include "share/atm_process/atmosphere_diagnostic.hpp"
 
+#include "ekat/ekat_pack.hpp"
```
[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/src/diagnostics/field_at_level.hpp:6]: Ah, yes. I'm a big fan of include-what-you-use), so I removed some includes in `field.hpp` which were not needed, and it caused compilation errors downstream (like in this diagnostic).

File: components/eamxx/tests/uncoupled/p3/input.yaml [hunk 4]
```diff
@@ -12,6 +12,7 @@ Time Stepping:
 atmosphere_processes:
   atm_procs_list: (p3)
   p3:
+    compute_tendencies: [T_mid,qc]
```
[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/tests/uncoupled/p3/input.yaml:15]: Yeah, true. I didn't want to test it b/c of the large number of fields to check, but it's true that cldfrac is a small process, so I might do it there.

File: components/eamxx/src/share/atm_process/atmosphere_process.cpp [hunk 5]
```diff
@@ -93,8 +93,39 @@ void AtmosphereProcess::run (const double dt) {
...
+        const auto& f     = get_field_out(fname);
+
+        auto& step_tend = m_proc_tendencies_single_step.at(tname);
+        auto& tend      = it.second;
+
+        step_tend.update(f,1/dt_sub,-1/dt_sub);
+        tend.update(step_tend,1.0,1.0);
+      }
+      stop_timer(m_timer_prefix + this->name() + "::compute_tendencies");
+    }
```
[the remark this answers is not in the document]
Comment (tcclevenger) [reply components/eamxx/src/share/atm_process/atmosphere_process.cpp:127]: I wonder if we even need it for those. I guess it just adds code since that's the only place it's called. 

At some point soonish I'll update the conservation checker and if I find something nicer to do for all of them I'll change it. So no, I don't feel strongly.

[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/src/share/atm_process/atmosphere_process.cpp:127]: Ok, done.

File: components/eamxx/tests/uncoupled/p3/input.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/tests/uncoupled/p3/input.yaml:15]: Actually, cldfrac does not update any var, it only _computes_ a var (I restricted tendencies to inout vars).

[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/tests/uncoupled/p3/input.yaml:15]: I'll do shoc, it has "only" 8 vars

**Comment by bartgol** [comment]
@jgfouca Any reason why the fake tests are run on this PR? I thought they would be disabled on normal testing, and run only on scripts testing...

```
Build type full_sp_debug failed at testing time. Here's a list of failed tests:
107:p3_baseline_f90_fake_omp16
108:p3_baseline_cxx_fake_omp16
143:shoc_baseline_f90_fake_omp16
144:shoc_baseline_cxx_fake_omp16
```

**Comment by jgfouca** [comment]
@bartgol , those tests didn't actually fail. test-all-scream just found them in the output.

**Comment by bartgol** [comment]
I'm quite puzzled by the SMS_D v1 fail. It fails for only one field, namely tke, which happens to be *zero* in the baselines (correct, since tke is not in the IC file, so tke(t=0)=0), but tke=0.0004 in the branch. I am not sure where this value comes from. tendencies calculation is also not enabled in any of our v1 tests (perhaps I should enable one, btw), so that's not the culprit.

**Comment by PeterCaldwell** [comment]
I think tke is only changed by SHOC and advected by the dycore. Maybe wet/dry isn't being done consistently? I'm grasping at straws... Hopefully we get a pass on this next round of testing.

**Comment by bartgol** [comment]
Ok, when I manually ran the SMS test I forgot to add `-c -b master`, so no baseline cmp was performed. Of course it passed. Adding those, I can reproduce the fail. I will debug a bit, to figure out why tke=0.0004 in the branch upon initialization.

**Comment by bartgol** [comment]
Ok, I think something is off in master, not this branch, although I don't know why master doesn't fail.

So, SHOC does this during initialization:
```
    if (run_type==RunType::Initial){
      Kokkos::deep_copy(sgs_buoy_flux,0.0);
      Kokkos::deep_copy(tk,0.0);
      Kokkos::deep_copy(tke,0.0004);
      Kokkos::deep_copy(tke_copy,0.0004);
      Kokkos::deep_copy(cldfrac_liq,0.0);
    }
```
So it makes sense that I see tke=0.0004 saved to the output file for t=0. However, master's baselines have tke=0 at t=0. Most importantly, I don't know why the nightlies don't fail for that test. I'm gonna run master and check that it indeed generates zero's (rather than, say, generate 0.0004, but somehow pass the check against master baselines).

**Comment by bartgol** [comment]
@AaronDonahue I found why I get a diff, and it is expected.

**tl; dr**: we can merge and bless.

The culprit is [this commit](https://github.com/E3SM-Project/scream/pull/2216/commits/b1485a4e64f6c08210003c950a3f35751b080745). The commit moved the OutputManagers init till _after_ atm procs init. This means that any initialization that happens in atm procs was not reflected in the t=0 output in master. This PR changes this, and any field that is initialized by an atm proc during init, will see that initialization reflected in the t=0 output. In particular, shoc inits "tke" to 0.0004 during initialization. This was not reflected in t=0 output in master, but it is reflected in t=0 output in this branch.

I think the way this branch works is more correct, since t=0 output should reflect the state of each field at the end of initialization sequence.

File: components/eamxx/tests/uncoupled/shoc/CMakeLists.txt [hunk 6]
```diff
@@ -34,8 +34,8 @@ if (TEST_RANK_END GREATER TEST_RANK_START)
   math (EXPR CMP_RANK_START ${TEST_RANK_START}+1)
   foreach (MPI_RANKS RANGE ${CMP_RANK_START} ${TEST_RANK_END})
 
-    set (SRC_FILE "${BASE_TEST_NAME}_standalone_output.INSTANT.nsteps_x${NUM_STEPS}.np${MPI_RANKS}.nc")
-    set (TGT_FILE "${BASE_TEST_NAME}_standalone_output.INSTANT.nsteps_x${NUM_STEPS}.np${TEST_RANK_START}.nc")
+    set (SRC_FILE "${BASE_TEST_NAME}_standalone_output.INSTANT.nsteps_x1.np${MPI_RANKS}.nc")
```
Comment (AaronDonahue) [inline components/eamxx/tests/uncoupled/shoc/CMakeLists.txt:37]: Do all of these changes mean that NUM_STEPS is deprecated?  If so, why not test at least one of these runs with >1 steps?

**Review by AaronDonahue — APPROVED** [review]
Looks good to me.  I think we can go ahead and merge now that we understand the DIFF that we were seeing.
